R/helper_patient_data_fix.R

```diff
@@ -172,27 +172,89 @@ parse_dates <- function(date) {
 
 #' @title Check value against list of allowed values.
 #'
+#' @description
+#' Invalid values are logged and depending on the replace invalid flag
+#' also replaced with an error value. When checking for the validity of values
+#' the case of the string is not considered nor are spaces and non-alphanumeric
+#' characters. If the value matches an entry in the allowed_values list
+#' that match is returned as the result.
+#'
 #' @param x character value to check.
 #' @param valid_values list of valid options.
-#' @param error_val value that is used to signal invalid values.
 #' @param id patient id.
+#' @param replace_invalid A flag controlling whether to also replace invalid .
+#' @param error_val value that is used to signal invalid values.
 #' @param col column name when used with mutate/across.
 #'
 #' @return Either x or "Other".
-check_allowed_values <- function(x, valid_values, error_val, id, col = "") {
+check_allowed_values <- function(x, valid_values, id, replace_invalid = TRUE, error_val = "Undefined", col = "") {
     if (is.na(x) || x == "") {
         return(NA_character_)
     }
 
-    if (!tolower(x) %in% tolower(valid_values)) {
+    valid_value_mapping <- setNames(as.list(valid_values), sanitize_str(valid_values))
+
+    if (!sanitize_str(x) %in% names(valid_value_mapping)) {
         logWarn("Patient ", id, ": Value ", x, " for column ", col, " is not in the list of allowed values. ")
-        if (!is.na(error_val)) {
+        if (replace_invalid) {
             logInfo("Replacing ", x, " with ", error_val, ".")
-            x <- error_val
+            return(error_val)
```

Comment:
instead of two return statements at the end, we could just return x as last statement (or not at all if the if command is returning a value like in other programming languages, not sure about R)

Reply:
I'm not sure how to improve the overall control flow here. Right now I find the readability ok as the return statements are closely related to the if statements. My impression is that this would become less clear if I try to move return statements around. In the end the function is already a little bit too complex with three different return routes. I would leave it as it is we see a very concrete readability improvement.

Reply:
Now it is more obvious as there is another function called